rpcs3/Emu/Cell/SPUThread.cpp

```diff
@@ -1189,28 +1168,33 @@ bool SPUThread::get_ch_value(u32 ch, u32& out)
 			return true;
 		}
 
-		vm::waiter waiter;
-
+		vm::waiter* waiter = nullptr;
 		if (ch_event_mask & SPU_EVENT_LR)
 		{
-			waiter.owner = this;
-			waiter.addr = raddr;
-			waiter.size = 128;
-			waiter.stamp = rtime;
-			waiter.data = rdata.data();
-			waiter.init();
+			waiter = new vm::waiter();
+			waiter->owner = this;
+			waiter->addr = raddr;
+			waiter->stamp = rtime;
+			waiter->data = rdata.data();
+			waiter->init();
 		}
 
 		while (!(res = get_events(true)))
 		{
 			if (test(state & cpu_flag::stop))
 			{
+				waiter->remove();
 				return false;
 			}
 
 			thread_ctrl::wait_for(100);
 		}
 
+		if (waiter != nullptr)
+		{
+			waiter->remove();
```

Comment:
Does waiter remove really delete self?

Reply:
Yes, queues it for delete so the removal won't block the continuation of execution. It can be put in a wrapper object with a destructor calling remove, to make it more clear